File: tools/vim-vize/run-real-server.mjs
// Drive the packaged Vim integration through a pinned vim-lsp checkout and a
// real `vize lsp` process. All response assertions live in the Vim archive's
// own test directory so downstream users can run the same host scenario.
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import {
  prepareRealVueWorkspace,
  repositoryRoot,
  resolveRealServerPath,
} from "../editor-e2e/real-vue-workspace.mjs";

const vimPath = process.env.VIZE_TEST_VIM_PATH?.trim() || "vim";
const vimLspPath = process.env.VIZE_TEST_VIM_LSP_PATH?.trim();
if (!vimLspPath) {
  throw new Error(
    "VIZE_TEST_VIM_LSP_PATH must point at the pinned vim-lsp checkout used by the host test",
  );
}

const vimLspPlugin = path.join(vimLspPath, "plugin", "lsp.vim");
if (!fs.existsSync(vimLspPlugin)) {
  throw new Error(`vim-lsp plugin not found: ${vimLspPlugin}`);
}

const pluginRoot = path.join(repositoryRoot, "editors", "vim");
const specPath = path.join(pluginRoot, "test", "vize_e2e_spec.vim");
const serverPath = resolveRealServerPath();
const sessionPath = fs.mkdtempSync(path.join(os.tmpdir(), "vize-vim-e2e-"));
const workspacePath = path.join(sessionPath, "real-vue");
const errorPath = path.join(sessionPath, "vim-errors.log");
const verbosePath = path.join(sessionPath, "vim-verbose.log");

prepareRealVueWorkspace(workspacePath);

try {
  const result = spawnSync(
    vimPath,
    ["-Nu", "NONE", "-n", "-es", "-i", "NONE", `-V1${verbosePath}`, "-S", specPath],
    {
      cwd: workspacePath,
      encoding: "utf-8",
      env: {
        ...process.env,
        VIZE_E2E_ERROR_PATH: errorPath,
        VIZE_E2E_PLUGIN_ROOT: pluginRoot,
        VIZE_E2E_SERVER: serverPath,
        VIZE_E2E_WORKSPACE: workspacePath,
      },
      timeout: 1_200_000,
    },
  );

  if (result.stdout) process.stdout.write(result.stdout);
  if (result.stderr) process.stderr.write(result.stderr);

  if (result.error) {
    throw result.error;
  }
  if (result.status !== 0) {
    for (const diagnosticPath of [errorPath, verbosePath]) {
      if (fs.existsSync(diagnosticPath)) {
        process.stderr.write(fs.readFileSync(diagnosticPath, "utf-8"));
      }
    }
    throw new Error(`headless Vim scenario failed with exit code ${result.status}`);
  }
  console.log("vim real-server scenario passed");
} finally {
  fs.rmSync(sessionPath, { force: true, recursive: true });
}

File: tools/commands/editors/vim/run-real-server.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [package]
//! edition = "2024"
//!
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! ```

use std::{
    env, fs,
    path::PathBuf,
    process::{Command, ExitCode, Stdio},
};

#[path = "../../../rust/common.rs"]
mod common;
#[path = "../../../rust/editor_e2e.rs"]
mod editor_e2e;

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let repo = common::repo_root()?;
    let vim = env::var("VIZE_TEST_VIM_PATH").unwrap_or_else(|_| "vim".to_string());
    let vim_lsp = env::var("VIZE_TEST_VIM_LSP_PATH").map_err(|_| {
        "VIZE_TEST_VIM_LSP_PATH must point at the pinned vim-lsp checkout used by the host test"
            .to_string()
    })?;
    let vim_lsp_plugin = PathBuf::from(&vim_lsp).join("plugin/lsp.vim");
    if !vim_lsp_plugin.exists() {
        return Err(format!(
            "vim-lsp plugin not found: {}",
            vim_lsp_plugin.display()
        ));
    }
    let plugin_root = repo.join("editors/vim");
    let spec_path = plugin_root.join("test/vize_e2e_spec.vim");
    let server_path = editor_e2e::resolve_real_server_path(&repo)?;
    let session = unique_temp_dir("vize-vim-e2e")?;
    let workspace = session.join("real-vue");
    let error_path = session.join("vim-errors.log");
    let verbose_path = session.join("vim-verbose.log");
    editor_e2e::prepare_real_vue_workspace(&workspace, false)?;
    let result = Command::new(vim)
        .arg("-Nu")
        .arg("NONE")
        .arg("-n")
        .arg("-es")
        .arg("-i")
        .arg("NONE")
        .arg(format!("-V1{}", verbose_path.display()))
        .arg("-S")
        .arg(&spec_path)
        .current_dir(&workspace)
        .env("VIZE_E2E_ERROR_PATH", &error_path)
        .env("VIZE_E2E_PLUGIN_ROOT", &plugin_root)
        .env("VIZE_E2E_SERVER", &server_path)
        .env("VIZE_E2E_WORKSPACE", &workspace)
        .stdin(Stdio::null())
        .output();
    let output = result.map_err(|error| format!("failed to run headless Vim: {error}"))?;
    print!("{}", String::from_utf8_lossy(&output.stdout));
    eprint!("{}", String::from_utf8_lossy(&output.stderr));
    if !output.status.success() {
        for path in [&error_path, &verbose_path] {
            if path.exists() {
                eprint!("{}", common::read_text(path)?);
            }
        }
        let _ = fs::remove_dir_all(&session);
        return Err(format!(
            "headless Vim scenario failed with exit code {}",
            output.status.code().unwrap_or(1)
        ));
    }
    let _ = fs::remove_dir_all(&session);
    println!("vim real-server scenario passed");
    Ok(())
}

fn unique_temp_dir(prefix: &str) -> Result<PathBuf, String> {
    let path = env::temp_dir().join(format!(
        "{}-{}-{}",
        prefix,
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map_err(|error| error.to_string())?
            .as_nanos()
    ));
    common::mkdir(&path)?;
    Ok(path)
}
